### DISCUSSION TITLE
Observability

### DISCUSSION THREAD

**Discussion by AngusWaller** [eclipse-zenoh/roadmap#92] [Ideas and new features]
A key aspect of software being adopted by industry is observability. Looking through the Issues & Discussions it seems like this key aspect has been raised [here](https://github.com/eclipse-zenoh/zenoh/issues/48) however, there has been no discussion or movement for this it appears. In quite a few blog posts Zenoh is compared to products like Kafka, so it would be good to have the same level of observability as these products.

Observability in this context would be tracking of for example
* Zenoh router health, memory consumption, CPU consumption, ingress messages, egress messages
* Message metrics such as producer/consumer delay
* Storage metrics such as storage queueing
* Replication metrics such as connection health, strong eventual consistency metrics
* Any metrics that would help debug an issue with Zenoh
* Any metrics that would help debug an issue with Zenoh interfacing with external systems

Using similar design patterns to Zenoh Storage of using multiple supported backends I believe it would be possible to create multiple implementations for metrics & logging, however, another approach may be to use OpenTelemetry and expect observability products to parse from OpenTelemetry.

This will impact the speeds achievable by Zenoh which may hamper the primary value proposition somewhat, however, if this was configurable and opt-in similar to storage, it would allow consumers the flexibility of determining when/where to observe. 

Happy to contribute! Would like input from Zenoh authors' first

**Comment by kydos** [comment]
Hello @AngusWaller,  observability is indeed a key aspect of infrastructure software. 

Some  informations are available today as part of the admin space. That said we are going through a roadmap exercise and observability  is one of the area of focus for the next release(s). 

Are you part of the [Zenoh Discord Channel](https://discord.gg/cRFZDJfS3g)? We will announce on the forum some open discussion when the community may be able to pitch-in. Otherwise, we'll be sharing our plan and wish-lists in the weeks to come.
